### ISSUE TITLE
parser.js breaks under minification

### ISSUE THREAD

**Issue by Alxandr** [aurelia/binding#13]
Parser.js breaks under minification. I have yet to be able to figure out exactly what goes wrong, but it's in the constructor of the `Parser` class. 

Here is the minified code that goes wrong
```js
function e(){this.cache={},this.lexer=new t}
```

The error that get's throw is

```
TypeError: Cannot set property 'cache' of undefined
```

Will do some more investigating.

**Comment by EisenbergEffect** [comment]
Any more insight into this one?

**Comment by martingust** [comment]
This issue seems to be because of this line: https://github.com/aurelia/dependency-injection/blob/master/src/util.js#L16

Class names gets mangled to lower case and we can't distinguish between class and function by looking at casing.

**Comment by EisenbergEffect** [comment]
Of course! That makes perfect sense. So, perhaps we should assume that everything is a class unless it is registered as a function. So we can add a new api `registerFunction` or `registerFactory` to let someone explicitly say that it's not a class. Thoughts? Opinions?

**Comment by Alxandr** [comment]
@EisenbergEffect Rather, use metadata to have it be a function, imho.

**Comment by Alxandr** [comment]
(with annotations):

```javascript
@Factory
function foo(bar) {
  return bar + baz;
}

di.register(foo, foo);
```

**Comment by EisenbergEffect** [comment]
Bingo! That's it. @Alxandr  I'll get this done before this week's release.

**Comment by martingust** [comment]
Yes. +1 for metadata!

**Comment by martingust** [comment]
Also this will be an issue when Custom Elements names are mangled: https://github.com/aurelia/templating/blob/master/src/resource-registry.js#L11

Should we add metadata here as well to set name? Would be nice to not have to though.

Other than these two issues Skeleton is loading fine minified.

**Comment by Alxandr** [comment]
In general, @EisenbergEffect, names **should not** be used.

**Comment by EisenbergEffect** [comment]
@martingust @Alxandr The new version of the module analyzer does not use function names for resources. That was an accident. It now correctly uses Export names, which should not be mangled by a minifier (otherwise that would break the API). You can, and have always been able to, specify the names of all resources. The name is only used for conventions, and in the next release is properly based on module export name, not function name.

**Comment by martingust** [comment]
When testing further I'm getting ViewEngine undefined on this line: https://github.com/aurelia/framework/blob/master/src/aurelia.js#L47

Haven't been able to dig into why, hopefully will have time later today.

**Comment by EisenbergEffect** [comment]
Fixed in 667a16e4dbd07b75493e0690ac4232d8c0cb508c
